### PR TITLE
Added team rate command

### PR DESCRIPTION
**Pull request by Chillee**
Added a team rate command.

You can test it out now on the server by doing `^teamrate chilli !okarinn` as well as `^howmanyfortourist`

### CODE REVIEW

File: tle/__main__.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def main():
 
     setup()
 
-    bot = commands.Bot(command_prefix=commands.when_mentioned_or(';'))
+    bot = commands.Bot(command_prefix=commands.when_mentioned_or('^'))
```
Comment (krofna) [inline tle/__main__.py]: revert this change

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "editor.formatOnSave": false
```
Comment (krofna) [inline .vscode/settings.json]: Should this file be in repo?

File: tle/cogs/codeforces.py [hunk 3]
```diff
@@ -244,6 +246,84 @@ def check(problem):
...
+                cur_cnt += step
+        cur_cnt += 1
+        mxRating = Codeforces.composeRatings(ratings*cur_cnt)
+        print(mxRating)
+        if mxRating < tourist_rating:
+            await ctx.send(f"Not even {1<<15} of {handle[0]} could beat tourist <:tourist_mad:556968281982894080>")
+        elif cur_cnt == 1:
+            await ctx.send(f"Tourist is already no match for {handle[0]} <:tourist_think:556968318909808682>")
+        else:
+            await ctx.send(f"With {cur_cnt} copies of {handle[0]}, {handle[0]} could beat tourist! Achieving a rating of {mxRating}. <:tourist:556976108462145541>")
```
Comment (krofna) [inline tle/cogs/codeforces.py]: maybe the emojis should be constants rather than inlined here

File: tle/cogs/codeforces.py [hunk 4]
```diff
@@ -244,6 +246,84 @@ def check(problem):
...
+            ratings = [user.rating for user in users if user.rating]
+        if len(ratings) == 0:
+            await ctx.send("No CF usernames with ratings passed in :'(")
+            return
+
+        teamRating = Codeforces.composeRatings(ratings)
+        if is_entire_server:
+            await ctx.send(f"The entire server's team rating is {teamRating}!")
+        else:
+            await ctx.send(f'The team rating is {teamRating}!')
```
[the remark this answers is not in the document]
Comment (meooow25) [reply tle/cogs/codeforces.py]: Agree with this.

File: .vscode/settings.json (hunk 2 above)
Comment (meooow25) [reply .vscode/settings.json]: You can add it to gitignore if you plan to keep using vscode for development.

File: tle/cogs/codeforces.py (hunk 3 above)
Comment (meooow25) [reply tle/cogs/codeforces.py]: I suggest putting the custom emojis we want to use in `discord_common.py`.

File: tle/cogs/codeforces.py [hunk 5]
```diff
@@ -244,6 +246,84 @@ def check(problem):
...
+            r = (left + right) / 2.0
+
+            rWinsProbability = 1.0
+            for rating in ratings:
+                rWinsProbability *= Codeforces.getEloWinProbability(r, rating)
+
+            if rWinsProbability==0:
+                left = r
+                continue
+            rating = log10(1 / (rWinsProbability) - 1) * 400 + r
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: Import `math`, and use `math.log10` instead.

File: tle/cogs/codeforces.py [hunk 6]
```diff
@@ -244,6 +246,84 @@ def check(problem):
             embed = discord.Embed(title=contest.name, url=contest.url)
             await ctx.send(f'Recommended contest for `{str_handles}`', embed=embed)
 
+    @staticmethod
+    def getEloWinProbability(ra: float, rb: float) -> float:
+        return 1.0 / (1 + 10**((rb - ra) / 400.0))
+
+    @staticmethod
+    def composeRatings(ratings: List[float]) -> int:
+        left = 100.0
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: `left` should be lower since we can have users with negative rating (current lowest is -47).

File: tle/cogs/codeforces.py [hunk 7]
```diff
@@ -244,6 +246,84 @@ def check(problem):
...
+    @commands.command(brief='Calculate team rating')
+    async def teamrate(self, ctx, *handles: str):
+        handles = handles or ('!' + str(ctx.author),)
+        is_entire_server = (handles[0] == 'all' and len(handles) == 1)
+        if is_entire_server:
+            res = cf_common.user_db.getallhandleswithrating()
+            ratings = [rating for _, _, rating in res]
+        else:
+            handles = await cf_common.resolve_handles(ctx, self.converter, handles, mincnt=1, maxcnt=1000)
+            users = await cf.user.info(handles=handles)
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: The call to the API can fail, which should be handled. We need a better way, but for now a `RunHandleCoroFailedError` can be raised which will be handled in the cog's error handler (see [here](https://github.com/cheran-senthil/TLE/blob/master/tle/cogs/handles.py#L131) for an example).

File: tle/cogs/codeforces.py [hunk 8]
```diff
@@ -244,6 +246,84 @@ def check(problem):
...
+            if rating > r:
+                left = r
+            else:
+                right = r
+        return round((left + right) / 2)
+
+    @commands.command(brief='Calculate team rating')
+    async def teamrate(self, ctx, *handles: str):
+        handles = handles or ('!' + str(ctx.author),)
+        is_entire_server = (handles[0] == 'all' and len(handles) == 1)
```
Comment (meooow25) [inline tle/cogs/codeforces.py]: Using `all` is questionable since all is an actual handle. I recommend using `+server` instead.
Also it can be simplified to `is_entire_server = handles == ('+server',)`

File: .vscode/settings.json (hunk 2 above)
Comment (Chillee) [reply .vscode/settings.json]: Hmm, I don't think it's unreasonable to have a `.vscode` file checked in for development purposes/consistency.

I'm fine with removing it though.

**Comment by krofna** [comment]
What happened to this pull request?

**Comment by cheran-senthil** [comment]
well if the changes weren't made, we shouldn't merge, no?

File: .vscode/settings.json (hunk 2 above)
Comment (cheran-senthil) [reply .vscode/settings.json]: Removed

**Comment by krofna** [comment]
waturr

**Comment by Chillee** [comment]
Alright I fixed all the issues mentioned - should be ready for review.

File: tle/cogs/codeforces.py [hunk 9]
```diff
@@ -442,6 +445,78 @@ def make_page(chunk):
...
+            for a, b in handle_map.items():
+                if b > 1:
+                    user_strs.append(f'{original_to_cf[a]}*{b}')
+                elif b == 1:
+                    user_strs.append(original_to_cf[a])
+                elif b < 0:
+                    raise CodeforcesCogError('How can you have negative members in team?')
+
+            user_str = ', '.join(user_strs)
+            ratings = [user.rating for user in users if user.rating for _ in range(handle_map[cf_to_original[user.handle.lower()]])]
```
Comment (krofna) [inline tle/cogs/codeforces.py:509]: cf usernames can have uppercase characters so this is broken

File: tle/cogs/codeforces.py [hunk 10]
```diff
@@ -442,6 +445,78 @@ def make_page(chunk):
...
+            res = cf_common.user_db.get_cf_users_for_guild(ctx.guild.id)
+            ratings = [cf_user.rating for user_id, cf_user in res if cf_user.rating is not None]
+            user_str = '+server'
+        else:
+            handle_map = {}
+            parsed_handles = []
+            for i in handles:
+                parse_str = i.split('*')
+                if len(parse_str) > 1:
+                    handle_map[parse_str[0].lower()] = int(parse_str[1])
```
Comment (krofna) [inline tle/cogs/codeforces.py:489]: raise exception if this isn't an int

File: tle/cogs/codeforces.py [hunk 11]
```diff
@@ -442,6 +445,84 @@ def make_page(chunk):
...
+        handles = handles or ('!' + str(ctx.author),)
+        is_entire_server = (handles == ('+server',))
+        if is_entire_server:
+            res = cf_common.user_db.get_cf_users_for_guild(ctx.guild.id)
+            ratings = [cf_user.rating for user_id, cf_user in res if cf_user.rating is not None]
+            user_str = '+server'
+        else:
+            def normalize(x):
+                return [i.lower() for i in x]
+            handle_map = {}
```
Comment (algmyr) [inline tle/cogs/codeforces.py:486]: Very generic name, maybe do `handle_counts` or similar

File: tle/cogs/codeforces.py [hunk 12]
```diff
@@ -442,6 +445,84 @@ def make_page(chunk):
...
+                    raise CodeforcesCogError('How can you have negative members in team?')
+
+            user_str = ', '.join(user_strs)
+            ratings = [user.rating for user in users if user.rating for _ in range(handle_map[cf_to_original[user.handle.lower()]])]
+
+        if len(ratings) == 0:
+            raise CodeforcesCogError("No CF usernames with ratings passed in.")
+
+        left = -100.0
+        right = 5000.0
```
Comment (algmyr) [inline tle/cogs/codeforces.py:521]: With the user multiplicity added 5000 is not high enough. MiFaFaOvO*100000 have combined rating 5744. You could just go overkill and do 10000.
